File: ios/MoodleLMS/Features/CourseDetail/CourseDetailRepository.swift
import Foundation

struct Section: Identifiable, Hashable {
    let id: Int
    let name: String
    let modules: [Module]
}

struct Module: Identifiable, Hashable {
    let id: Int
    let name: String
    let url: URL?
    let iconURL: URL?
    let modName: String?
}

final class CourseDetailRepository {
    private let client: MoodleClient
    private let tokenStore: KeychainStore

    init(client: MoodleClient, tokenStore: KeychainStore) {
        self.client = client
        self.tokenStore = tokenStore
    }

    func loadSections(courseId: Int) async throws -> [Section] {
        let dtos: [SectionDTO] = try await client.callFunction(
            "core_course_get_contents",
            params: ["courseid": String(courseId)]
        )
        let token = tokenStore.token()
        return dtos.map { dto in
            let cleaned = dto.name.htmlUnescaped
            return Section(
                id: dto.id,
                name: cleaned.isEmpty ? "Untitled section" : cleaned,
                modules: (dto.modules ?? []).map { m in
                    Module(
                        id: m.id,
                        name: m.name.htmlUnescaped,
                        url: m.url.flatMap(URL.init(string:)),
                        iconURL: m.modicon.flatMap { Self.attach(token: token, raw: $0) },
                        modName: m.modname
                    )
                }
            )
        }
    }

    private static func attach(token: String?, raw: String) -> URL? {
        guard let token, !token.isEmpty else { return URL(string: raw) }
        let separator = raw.contains("?") ? "&" : "?"
        return URL(string: "\(raw)\(separator)token=\(token)")
    }
}

File: android/app/src/main/java/io/itcorner/moodle/feature/coursedetail/CourseDetailRepository.kt
package io.itcorner.moodle.feature.coursedetail

import io.itcorner.moodle.core.auth.TokenStore
import io.itcorner.moodle.core.network.MoodleApi
import io.itcorner.moodle.core.network.MoodleApiException
import io.itcorner.moodle.feature.courses.htmlUnescape
import java.io.IOException
import javax.inject.Inject
import javax.inject.Singleton

data class Section(
    val id: Long,
    val name: String,
    val visible: Boolean,
    val modules: List<Module>,
)

data class Module(
    val id: Long,
    val name: String,
    val url: String?,
    val iconUrl: String?,
    val modName: String?,
)

@Singleton
class CourseDetailRepository @Inject constructor(
    private val api: MoodleApi,
    private val tokenStore: TokenStore,
) {
    suspend fun loadSections(courseId: Long): Result<List<Section>> = try {
        val token = tokenStore.token()
        val sections = api.getCourseContents(courseId).map { dto ->
            Section(
                id = dto.id,
                name = dto.name.htmlUnescape().ifBlank { "Untitled section" },
                visible = (dto.visible ?: 1) == 1,
                modules = dto.modules.map { m ->
                    Module(
                        id = m.id,
                        name = m.name.htmlUnescape(),
                        url = m.url,
                        iconUrl = m.modicon?.let { withToken(it, token) },
                        modName = m.modname,
                    )
                },
            )
        }
        Result.success(sections)
    } catch (e: IOException) {
        Result.failure(e)
    } catch (e: MoodleApiException) {
        Result.failure(e)
    }

    private fun withToken(url: String, token: String?): String {
        if (token.isNullOrBlank()) return url
        val sep = if (url.contains("?")) "&" else "?"
        return "$url${sep}token=$token"
    }
}
